odoo/addons/test_lint/tests/test_docstring.py

```diff
@@ -0,0 +1,245 @@
+                        raise ValueError(e)
+
+            elif kind == 'returns':
+                if name:
+                    e = f'invalid ":{kind} {name}:", did you mean ":rtype: {name}"?'
+                    raise ValueError(e)
+
+            elif kind == 'type':
+                if not name:
+                    e = "empty :type:"
+                    raise ValueError(e)
+                params[name] = None
+                try:
+                    type_type = field_body.children[0].astext().strip()
+                except IndexError as exc:
+                    if name:
+                        e = f'invalid ":{kind} {name}:", did you mean ":{kind}: {name}"?'
+                        raise ValueError(e) from exc
+                    raise
+                type_param = types.setdefault(name, type_type)
+                if type_param != type_type:
+                    e = f"conflicting type: {type_param} (:param:) vs {type_type} (:type:)"
+                    raise ValueError(e)
+
+            elif kind == 'rtype':
+                try:
+                    rtype = field_body.children[0].astext().strip()
+                except IndexError as exc:
+                    if name:
+                        e = f'invalid ":{kind} {name}:", did you mean ":{kind}: {name}"?'
+                        raise ValueError(e) from exc
+                    raise
+
+    return list(params), types, rtype
+
+
+@tagged('-at_install', 'post_install')
+class TestDocstring(BaseCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
+        # The docstrings can use many more roles and directives than the
+        # one present natively in docutils. That's because we use Sphinx
+        # to render them in the documentation, and Sphinx defines the
+        # "Python Domain", a set of additional rules and directive to
+        # understand the python language.
+        #
+        # It is not desirable to add a dependency on Sphinx in
+        # community, at least not only for this linter.
+        #
+        # The following code adds a bunch of dummy elements for the
+        # missing roles and directives, so docutils is able to parse
+        # them with no warning.
+
+        def role_function(name, rawtext, text, lineno, inliner, options=None, content=None):
+            return [docutils.nodes.inline(rawtext, text)], []
+
+        for role in ('attr', 'class', 'func', 'meth', 'ref', 'const', 'samp', 'term'):
+            docutils.parsers.rst.roles.register_local_role(role, role_function)
+
+        for directive in ('attribute', 'deprecated'):
+            docutils.parsers.rst.directives.register_directive(
+                directive, docutils.parsers.rst.directives.admonitions.Note)
+
+    def test_docstring(self):
+        """ Verify that the function signature and its docstring match. """
+        registry = Registry(get_db_name())
+        seen_methods = set()
+
+        for model_name, model_cls in registry.items():
```

Comment:
So `l10n` modules won't be part of the test/docs because they are excluded from the CI, no?

Reply:
Wouldn't it be possible to have static code analysis (AKA a linter) so that we don't need to install modules?

Reply:
Not really because existing static analysis tools don't know about odoo inheritance. Right now, it's not a priority to have an odoo-specific linter, but maybe the LSP could one day cover such cases. Until then, this is the best compromise that we can do now.